File: A_ros/src/turtle_follower/scripts/state_publisher.py
#! /usr/bin/python

import sys
from math import sin, cos

import rospy
import tf2_ros
from rospy import loginfo as log
from turtlesim.msg import Pose

BASE_FRAME = 'world'
MAP_FRAME = 'map'
TURTLE_FRAME = lambda turtle: f'turtle_{turtle}'

broadcaster: tf2_ros.TransformBroadcaster = None


def send_map_transform(x=5.5, y=5.5, z=0, w=1):  # rename args
    assert z * z + w * w == 1

    static_br = tf2_ros.StaticTransformBroadcaster()
    map_t = tf2_ros.TransformStamped()
    map_t.header.stamp = rospy.Time.now()
    map_t.header.frame_id = BASE_FRAME
    map_t.child_frame_id = MAP_FRAME

    map_t.transform.translation.x = x
    map_t.transform.translation.y = y

    map_t.transform.rotation.z = z
    map_t.transform.rotation.w = w

    static_br.sendTransform(map_t)


def publish_transform(pose: Pose, t: tf2_ros.TransformStamped):
    t.header.stamp = rospy.Time.now()
    t.transform.translation.x = pose.x
    t.transform.translation.y = pose.y
    t.transform.rotation.z = sin(pose.theta / 2)
    t.transform.rotation.w = cos(pose.theta / 2)

    global broadcaster
    broadcaster.sendTransform(t)


def main():
    argv = rospy.myargv(argv=sys.argv)[1:]

    rospy.init_node("state_publisher")

    log('State publisher started')

    send_map_transform()
    log('Sent map frame.')

    global broadcaster
    broadcaster = tf2_ros.TransformBroadcaster()

    for turtle in argv:
        log(f'Started publishing transform for  {TURTLE_FRAME(turtle)}.')

        t = tf2_ros.TransformStamped()
        t.header.frame_id = BASE_FRAME
        t.child_frame_id = TURTLE_FRAME(turtle)
        t.transform.translation.z = 0
        t.transform.rotation.x = 0
        t.transform.rotation.y = 0

        rospy.Subscriber(f'/{turtle}/pose', Pose, publish_transform, t)

    rospy.spin()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSException:
        print('Failed.')

File: A_ros/src/turtle_follower_cpp/src/state_publisher.cc
#include <ros/ros.h>

#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/static_transform_broadcaster.h>

#include <geometry_msgs/TransformStamped.h>
#include <turtlesim/Pose.h>

#define WORLD_FRAME "world"
#define MAP_FRAME "map"

std::string TURTLE_FRAME(const std::string &name) {
    return "turtle_" + name;
}

std::string POSE_TOPIC(const std::string &name) {
    return "/" + name + "/pose";
}

#define MAP_X 5.5
#define MAP_Y 5.5

tf2_ros::TransformBroadcaster *broadcaster;

void callback(const turtlesim::Pose::ConstPtr &p, geometry_msgs::TransformStamped *t) {
    t->header.stamp = ros::Time::now();
    t->transform.translation.x = p->x;
    t->transform.translation.y = p->y;
    t->transform.rotation.z = sin(p->theta / 2);
    t->transform.rotation.w = cos(p->theta / 2);

    broadcaster->sendTransform(*t);
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "state_publisher");
    ros::NodeHandle node;
    ROS_INFO("Started...");

    //<editor-fold desc="Map Transform">
    geometry_msgs::TransformStamped trans;
    trans.child_frame_id = MAP_FRAME;

    trans.header.frame_id = WORLD_FRAME;
    trans.header.stamp = ros::Time::now();

    // Are they auto-inited to 0?
    trans.transform.translation.z = 0;
    trans.transform.rotation.x = 0;
    trans.transform.rotation.y = 0;

    trans.transform.translation.x = MAP_X;
    trans.transform.translation.y = MAP_Y;
    trans.transform.rotation.z = 0;
    trans.transform.rotation.w = 1;

    tf2_ros::StaticTransformBroadcaster staticBroadcaster;
    staticBroadcaster.sendTransform(trans);
    ROS_INFO("Map transform broadcasted.");
    //</editor-fold>


    tf2_ros::TransformBroadcaster br;
    broadcaster = &br;

    std::vector<ros::Subscriber> subs; // If i dont store them they apparently go away.
    std::vector<geometry_msgs::TransformStamped *> ts; // For deleting the Transforms

    for (++argv; *argv; ++argv) {
        if (*argv[0] == '_') continue; // Why isn't ros removing the '__log:=...' arg ?

        ROS_INFO("Publishing transform for the turtle: %s", *argv);

        // If i directly initialize object, in every iteration, it is stored in the same point in memory. !!!; thus use new and delete
        geometry_msgs::TransformStamped *t = new geometry_msgs::TransformStamped();
        t->child_frame_id = TURTLE_FRAME(*argv);
        t->header.frame_id = WORLD_FRAME;

        // Are they auto-inited to 0?
        t->transform.translation.z = 0;
        t->transform.rotation.x = 0;
        t->transform.rotation.y = 0;

        ts.push_back(t);

        // Or should i just use a class?
        // https://answers.ros.org/question/11810/how-to-pass-arguments-tofrom-subscriber-callback-functions/
        const boost::function<void(const turtlesim::Pose::ConstPtr &)> bound_callback = boost::bind(callback, _1, t);
        // https://answers.ros.org/question/12045/how-to-deliver-arguments-to-a-callback-function/
        subs.push_back(node.subscribe<turtlesim::Pose &>(POSE_TOPIC(*argv), 1, bound_callback));
    }

    ros::spin();

    for (geometry_msgs::TransformStamped *t: ts) delete t;
    return 0;
}